sys/kern/lockdep.c

```diff
@@ -233,30 +233,43 @@ void lockdep_init(void) {
 
 void lockdep_acquire(lock_class_mapping_t *lock) {
   lock_class_t *class;
-  held_lock_t *hlock;
+  thread_t *thread = thread_self();
 
   lockdep_lock();
 
-  if (thread_self()->td_lock_depth >= LOCKDEP_MAX_HELD_LOCKS)
+  if (thread->td_lock_depth >= LOCKDEP_MAX_HELD_LOCKS)
     panic("lockdep: max lock depth reached");
 
   if (!(class = lock->lock_class))
     class = get_or_create_class(lock);
 
-  hlock = &thread_self()->td_held_locks[thread_self()->td_lock_depth++];
-  hlock->lock_class = class;
+  thread->td_held_locks[thread->td_lock_depth++] = class;
 
   add_prev_link();
 
   lockdep_unlock();
 }
 
-void lockdep_release(void) {
+void lockdep_release(lock_class_mapping_t *lock) {
+  bool shift = false;
+  lock_class_t *class;
+  thread_t *thread = thread_self();
   lockdep_lock();
 
-  thread_self()->td_lock_depth--;
+  class = look_up_lock_class(lock);
+  assert(class);
+
+  /* Delete the class of the released lock and move the rest shift the rest */
+  for (int i = 0; i < thread->td_lock_depth; i++) {
+    if (class == thread->td_held_locks[i])
+      shift = true;
+    else if (shift)
+      thread->td_held_locks[i - 1] = thread->td_held_locks[i];
+  }
```

Comment:
What if we're holding two different locks of the same class? That could lead to the wrong class being removed when releasing one of them. I think we should restore the `held_lock_t` type and hold a pointer in it to the class as well as the exact lock that was acquired (or store just the lock, and access the class through it).

Reply:
But then we remember this class twice, so when we remove one we still have one left.

Reply:
Yes, but the exact pointer to the class being removed might not be the one that was added when the lock was acquired.

Reply:
Ok, I think I have convinced myself that this isn't an issue. This would be a problem if we could hold two locks belonging to the same class, with one lock from another class in between, e.g. `(A::lock1) (B::lock2) (A::lock3)`, but then acquiring `lock3` would create a cycle, so it's impossible.

Reply:
I also can't come up with a scenario where removing "not our" class could be considered harmful. In fact, when a new lock is being acquired the ordering of currently held locks doesn't matter.